packages/rspack/src/fileSystem.ts

```diff
@@ -12,33 +13,48 @@ function createThreadsafeNodeFSFromRaw(
 	fs: typeof import("fs")
 ): ThreadsafeWritableNodeFS {
 	return {
-		writeFile: (file, data) => fs.writeFileSync(file, data),
-		removeFile: file => fs.unlinkSync(file),
-		mkdir: dir => fs.mkdirSync(dir),
-		mkdirp: dir =>
-			fs.mkdirSync(dir, {
+		writeFile: (file, data) => {
+			return util.promisify(fs.writeFile.bind(fs))(file, data);
+		},
+		removeFile: file => {
+			return util.promisify(fs.unlink.bind(fs))(file);
+		},
+		mkdir: dir => {
+			return util.promisify(fs.mkdir.bind(fs))(dir);
+		},
+		mkdirp: dir => {
+			return util.promisify(fs.mkdir.bind(fs))(dir, {
 				recursive: true
-			}),
+			});
+		},
 		removeDirAll: dir => {
 			// memfs don't support rmSync
-			rmrfBuild(fs)(dir);
+			return rmrfBuild(fs)(dir);
 		}
 	};
 }
 
 const rmrfBuild = (fs: typeof import("fs")) => {
-	const rmrf = (dir: string) => {
-		if (fs.existsSync(dir)) {
-			const files = fs.readdirSync(dir);
-			files.forEach(file => {
+	async function exists(path: string) {
+		try {
+			await util.promisify(fs.access.bind(fs))(path);
+			return true;
+		} catch {
+			return false;
+		}
+	}
+	const rmrf = async (dir: string) => {
+		if (await exists(dir)) {
+			const files = await util.promisify(fs.readdir.bind(fs))(dir);
+			for (const file of files) {
```

Comment:
`Promise.all` is better

Reply:
Good catch, `output.clean = true` currently shows

![image](https://github.com/web-infra-dev/rspack/assets/1430279/3bf54b6c-5ee2-4168-9501-85d8322e36ed)